File: go/protocols/mpp/client/http_stream.go
// HTTP streaming helpers for metered sessions.
//
// LLM APIs commonly stream responses over Server-Sent Events (SSE) or chunked
// HTTP. This file keeps the parser transport-neutral (SseDecoder works on raw
// chunks from any reader), then layers a net/http-friendly stream and commit
// transport on top for applications that want batteries included.
//
// Behavior mirrors rust/crates/mpp/src/client/http_stream.rs; the TypeScript
// counterpart is typescript/packages/mpp/src/client/HttpStream.ts.
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/solana-foundation/pay-kit/go/protocols/mpp/intents"
)

// SseEvent is a parsed Server-Sent Event frame. Event, ID, and Retry are nil
// when the frame omitted the field.
//
// Mirrors rust SseEvent in rust/crates/mpp/src/client/http_stream.rs.
type SseEvent struct {
	Event *string
	Data  string
	ID    *string
	Retry *uint64
}

// SseDecoder is an incremental SSE decoder.
//
// Feed raw HTTP chunks with PushChunk. It returns all complete events decoded
// from that chunk and retains partial data internally.
//
// Mirrors rust SseDecoder in rust/crates/mpp/src/client/http_stream.rs.
type SseDecoder struct {
	buffer  string
	current SseEvent
}

// PushChunk decodes the events completed by a raw chunk of the stream body.
func (d *SseDecoder) PushChunk(chunk []byte) ([]SseEvent, error) {
	if !utf8.Valid(chunk) {
		return nil, fmt.Errorf("SSE chunk is not valid UTF-8")
	}
	d.buffer += string(chunk)

	var events []SseEvent
	for {
		index := strings.IndexByte(d.buffer, '\n')
		if index < 0 {
			break
		}
		line := d.buffer[:index]
		d.buffer = d.buffer[index+1:]
		line = strings.TrimSuffix(line, "\r")
		if event, ok := d.processLine(line); ok {
			events = append(events, event)
		}
	}
	return events, nil
}

// Finish flushes an incomplete final event, if any, at EOF.
func (d *SseDecoder) Finish() ([]SseEvent, error) {
	var events []SseEvent
	if d.buffer != "" {
		line := strings.TrimSuffix(d.buffer, "\r")
		d.buffer = ""
		if event, ok := d.processLine(line); ok {
			events = append(events, event)
		}
	}
	if event, ok := d.dispatchCurrent(); ok {
		events = append(events, event)
	}
	return events, nil
}

func (d *SseDecoder) processLine(line string) (SseEvent, bool) {
	if line == "" {
		return d.dispatchCurrent()
	}
	if strings.HasPrefix(line, ":") {
		return SseEvent{}, false
	}

	field := line
	value := ""
	if index := strings.IndexByte(line, ':'); index >= 0 {
		field = line[:index]
		value = strings.TrimPrefix(line[index+1:], " ")
	}

	switch field {
	case "event":
		event := value
		d.current.Event = &event
	case "data":
		if d.current.Data != "" {
			d.current.Data += "\n"
		}
		d.current.Data += value
	case "id":
		id := value
		d.current.ID = &id
	case "retry":
		if retry, err := strconv.ParseUint(value, 10, 64); err == nil {
			d.current.Retry = &retry
		}
	}
	return SseEvent{}, false
}

func (d *SseDecoder) dispatchCurrent() (SseEvent, bool) {
	if d.current.Event == nil && d.current.Data == "" && d.current.ID == nil && d.current.Retry == nil {
		return SseEvent{}, false
	}
	current := d.current
	d.current = SseEvent{}
	return current, true
}

// MeteredSseEventKind discriminates ParseMeteredSseEvent results.
type MeteredSseEventKind int

// MeteredSseEvent kinds, mirroring the rust MeteredSseEvent enum variants.
const (
	// MeteredSseEventMetering is an mpp.metering / metering directive event.
	MeteredSseEventMetering MeteredSseEventKind = iota

	// MeteredSseEventUsage is an mpp.usage / usage final-amount event.
	MeteredSseEventUsage

	// MeteredSseEventMessage is an application message event.
	MeteredSseEventMessage

	// MeteredSseEventDone is a done event or [DONE] sentinel message.
	MeteredSseEventDone

	// MeteredSseEventOther is an unrecognized event passed through untouched.
	MeteredSseEventOther
)

// MeteredSseEvent is a parsed metered SSE event. Exactly the field matching
// Kind is populated.
//
// Mirrors rust MeteredSseEvent in rust/crates/mpp/src/client/http_stream.rs.
type MeteredSseEvent struct {
	Kind     MeteredSseEventKind
	Metering *intents.MeteringDirective
	Usage    *intents.MeteringUsage
	Message  json.RawMessage
	Other    *SseEvent
}

// ParseMeteredSseEvent classifies an SSE event by the metered-session event
// names: "mpp.metering"/"metering", "mpp.usage"/"usage", "done", and the
// "[DONE]" sentinel on the default message event. Application messages keep
// their raw JSON payload for the caller to decode.
//
// Mirrors rust parse_metered_sse_event.
func ParseMeteredSseEvent(event SseEvent) (MeteredSseEvent, error) {
	eventName := "message"
	if event.Event != nil {
		eventName = *event.Event
	}
	switch eventName {
	case "mpp.metering", "metering":
		directive := intents.MeteringDirective{}
		if err := json.Unmarshal([]byte(event.Data), &directive); err != nil {
			return MeteredSseEvent{}, fmt.Errorf("invalid mpp.metering event: %w", err)
		}
		return MeteredSseEvent{Kind: MeteredSseEventMetering, Metering: &directive}, nil
	case "mpp.usage", "usage":
		usage := intents.MeteringUsage{}
		if err := json.Unmarshal([]byte(event.Data), &usage); err != nil {
			return MeteredSseEvent{}, fmt.Errorf("invalid mpp.usage event: %w", err)
		}
		return MeteredSseEvent{Kind: MeteredSseEventUsage, Usage: &usage}, nil
	case "done":
		return MeteredSseEvent{Kind: MeteredSseEventDone}, nil
	case "message":
		if strings.TrimSpace(event.Data) == "[DONE]" {
			return MeteredSseEvent{Kind: MeteredSseEventDone}, nil
		}
		if !json.Valid([]byte(event.Data)) {
			return MeteredSseEvent{}, fmt.Errorf("invalid SSE message event: %q", event.Data)
		}
		return MeteredSseEvent{Kind: MeteredSseEventMessage, Message: json.RawMessage(event.Data)}, nil
	default:
		other := event
		return MeteredSseEvent{Kind: MeteredSseEventOther, Other: &other}, nil
	}
}

// meteredStreamState pairs the live metering directive with the optional final
// usage amount.
//
// Mirrors rust MeteredStreamState.
type meteredStreamState struct {
	directive   *intents.MeteringDirective
	finalAmount *uint64
	done        bool
}

// applyEvent folds one SSE event into the state, returning the raw application
// message when the event carries one. A usage event must reference the live
// directive's deliveryId; it may override only the amount.
func (s *meteredStreamState) applyEvent(event SseEvent) (json.RawMessage, error) {
	parsed, err := ParseMeteredSseEvent(event)
	if err != nil {
		return nil, err
	}
	switch parsed.Kind {
	case MeteredSseEventMetering:
		s.directive = parsed.Metering
		return nil, nil
	case MeteredSseEventUsage:
		if s.directive != nil && parsed.Usage.DeliveryID != s.directive.DeliveryID {
			return nil, fmt.Errorf(
				"usage delivery %s does not match directive %s",
				parsed.Usage.DeliveryID, s.directive.DeliveryID)
		}
		amount, err := parsed.Usage.AmountBaseUnits()
		if err != nil {
			return nil, err
		}
		s.finalAmount = &amount
		return nil, nil
	case MeteredSseEventMessage:
		return parsed.Message, nil
	case MeteredSseEventDone:
		s.done = true
		return nil, nil
	default:
		return nil, nil
	}
}

// directiveForCommit returns the live directive with the final usage amount
// applied, erroring when the stream never emitted a metering event.
func (s *meteredStreamState) directiveForCommit() (intents.MeteringDirective, error) {
	if s.directive == nil {
		return intents.MeteringDirective{}, fmt.Errorf("stream did not include mpp.metering event")
	}
	directive := *s.directive
	if s.finalAmount != nil {
		directive.Amount = strconv.FormatUint(*s.finalAmount, 10)
	}
	return directive, nil
}

// MeteredSseSession is a transport-neutral state machine for one metered SSE
// stream: feed it decoded SSE events, then Ack to commit the final amount.
//
// Mirrors rust MeteredSseSession in
// rust/crates/mpp/src/client/http_stream.rs.
type MeteredSseSession struct {
	consumer *SessionConsumer
	state    meteredStreamState
}

// MeteredSse starts a metered SSE state machine borrowing this consumer.
//
// Mirrors rust SessionConsumer::metered_sse.
func (c *SessionConsumer) MeteredSse() *MeteredSseSession {
	return &MeteredSseSession{consumer: c}
}

// AcceptEvent folds one decoded SSE event into the stream state and returns
// the raw application message when the event carries one.
//
// Mirrors rust MeteredSseSession::accept_event.
func (s *MeteredSseSession) AcceptEvent(event SseEvent) (json.RawMessage, error) {
	return s.state.applyEvent(event)
}

// IsDone reports whether the stream signaled completion.
//
// Mirrors rust MeteredSseSession::is_done.
func (s *MeteredSseSession) IsDone() bool { return s.state.done }

// Ack commits the stream's final amount (the usage amount when reported,
// otherwise the directive's reserved amount) through the consumer.
//
// Mirrors rust MeteredSseSession::ack.
func (s *MeteredSseSession) Ack(ctx context.Context) (intents.CommitReceipt, error) {
	directive, err := s.state.directiveForCommit()
	if err != nil {
		return intents.CommitReceipt{}, err
	}
	return s.consumer.CommitDirective(ctx, directive)
}

// HTTPCommitTransport is a minimal net/http transport for commit endpoints.
// The zero value posts to each directive's CommitURL with the default client.
//
// Mirrors rust HttpCommitTransport in
// rust/crates/mpp/src/client/http_stream.rs.
type HTTPCommitTransport struct {
	// Client is the HTTP client. nil uses http.DefaultClient.
	Client *http.Client

	// DefaultCommitURL is the commit endpoint used when a directive omits
	// CommitURL.
	DefaultCommitURL string

	// Authorization is an optional Authorization header value attached to
	// every commit request.
	Authorization string
}

// Commit posts the payload as JSON to the directive's commit endpoint and
// decodes the receipt.
//
// Mirrors rust HttpCommitTransport::commit.
func (t *HTTPCommitTransport) Commit(
	ctx context.Context,
	directive intents.MeteringDirective,
	payload intents.CommitPayload,
) (intents.CommitReceipt, error) {
	url := t.DefaultCommitURL
	if directive.CommitURL != nil {
		url = *directive.CommitURL
	}
	if url == "" {
		return intents.CommitReceipt{}, fmt.Errorf("metering directive missing commitUrl")
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return intents.CommitReceipt{}, fmt.Errorf("encode commit payload: %w", err)
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return intents.CommitReceipt{}, fmt.Errorf("build commit request: %w", err)
	}
	request.Header.Set("Content-Type", "application/json")
	if t.Authorization != "" {
		request.Header.Set("Authorization", t.Authorization)
	}

	client := t.Client
	if client == nil {
		client = http.DefaultClient
	}
	response, err := client.Do(request)
	if err != nil {
		return intents.CommitReceipt{}, fmt.Errorf("commit request failed: %w", err)
	}
	defer func() { _ = response.Body.Close() }()

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		detail, _ := io.ReadAll(io.LimitReader(response.Body, 4096))
		return intents.CommitReceipt{}, fmt.Errorf(
			"commit endpoint returned %d: %s", response.StatusCode, string(detail))
	}

	receipt := intents.CommitReceipt{}
	if err := json.NewDecoder(response.Body).Decode(&receipt); err != nil {
		return intents.CommitReceipt{}, fmt.Errorf("invalid commit receipt: %w", err)
	}
	return receipt, nil
}

// MeteredSseStream reads a metered SSE response body, yielding raw application
// messages and committing the final amount on Ack.
//
// Mirrors rust ReqwestMeteredSseStream in
// rust/crates/mpp/src/client/http_stream.rs.
type MeteredSseStream struct {
	consumer *SessionConsumer
	body     io.Reader
	decoder  SseDecoder
	pending  []json.RawMessage
	state    meteredStreamState
	buf      []byte
}

// NewMeteredSseStream wraps a consumer and an SSE response body, e.g.
// http.Response.Body. The caller retains ownership of the body and closes it
// after the stream is drained.
//
// Mirrors rust ReqwestMeteredSseStream::new.
func NewMeteredSseStream(consumer *SessionConsumer, body io.Reader) *MeteredSseStream {
	return &MeteredSseStream{
		consumer: consumer,
		body:     body,
		buf:      make([]byte, 4096),
	}
}

// Next returns the next application message, or nil once the stream is done.
//
// Mirrors rust ReqwestMeteredSseStream::next.
func (s *MeteredSseStream) Next() (json.RawMessage, error) {
	for {
		if len(s.pending) > 0 {
			message := s.pending[0]
			s.pending = s.pending[1:]
			return message, nil
		}
		if s.state.done {
			return nil, nil
		}

		n, readErr := s.body.Read(s.buf)
		if n > 0 {
			events, err := s.decoder.PushChunk(s.buf[:n])
			if err != nil {
				return nil, err
			}
			if err := s.applyEvents(events); err != nil {
				return nil, err
			}
		}
		if readErr != nil {
			if readErr != io.EOF {
				return nil, fmt.Errorf("stream read failed: %w", readErr)
			}
			events, err := s.decoder.Finish()
			if err != nil {
				return nil, err
			}
			if err := s.applyEvents(events); err != nil {
				return nil, err
			}
			s.state.done = true
		}
	}
}

func (s *MeteredSseStream) applyEvents(events []SseEvent) error {
	for _, event := range events {
		message, err := s.state.applyEvent(event)
		if err != nil {
			return err
		}
		if message != nil {
			s.pending = append(s.pending, message)
		}
	}
	return nil
}

// Ack drains any remaining events and commits the stream's final amount.
//
// Mirrors rust ReqwestMeteredSseStream::ack.
func (s *MeteredSseStream) Ack(ctx context.Context) (intents.CommitReceipt, error) {
	if !s.state.done {
		for {
			message, err := s.Next()
			if err != nil {
				return intents.CommitReceipt{}, err
			}
			if message == nil {
				break
			}
		}
	}
	directive, err := s.state.directiveForCommit()
	if err != nil {
		return intents.CommitReceipt{}, err
	}
	return s.consumer.CommitDirective(ctx, directive)
}

// IntoConsumer returns the wrapped consumer for reuse on the next request.
//
// Mirrors rust ReqwestMeteredSseStream::into_consumer.
func (s *MeteredSseStream) IntoConsumer() *SessionConsumer {
	return s.consumer
}

File: python/src/solana_pay_kit/protocols/mpp/client/http_stream.py
"""HTTP streaming helpers for metered sessions.

LLM APIs commonly stream responses over Server-Sent Events (SSE) or chunked
HTTP. This module keeps the parser transport-neutral (an incremental
:class:`SseDecoder` plus the metered event state machine), then layers a small
httpx adapter on top for applications that want batteries included.

Two metering invariants are enforced while folding a stream: a usage event's
``deliveryId`` must match the live metering directive, and a usage event may
override only the committed amount, never the ``deliveryId``.
"""

from __future__ import annotations

import contextlib
import json
from collections.abc import Iterable, Iterator
from dataclasses import dataclass, replace
from typing import Any, Literal

from solana_pay_kit.protocols.mpp.client.session_consumer import SessionConsumer
from solana_pay_kit.protocols.mpp.intents.session import (
    CommitPayload,
    CommitReceipt,
    MeteringDirective,
    MeteringUsage,
)

__all__ = [
    "SseEvent",
    "SseDecoder",
    "MeteredSseEvent",
    "parse_metered_sse_event",
    "MeteredSseSession",
    "MeteredSseStream",
    "HttpCommitTransport",
]


@dataclass
class SseEvent:
    """A parsed Server-Sent Event frame."""

    #: Event name from the ``event:`` field, or ``None`` for a default message.
    event: str | None = None
    #: Concatenated ``data:`` field payload (lines joined with newlines).
    data: str = ""
    #: Last-event-id from the ``id:`` field, if present.
    id: str | None = None
    #: Reconnection delay in milliseconds from the ``retry:`` field, if present.
    retry: int | None = None


class SseDecoder:
    """Incremental SSE decoder.

    Feed raw HTTP chunks with :meth:`push_chunk`. It returns all complete
    events decoded from that chunk and retains partial data internally.
    """

    def __init__(self) -> None:
        """Create an empty decoder."""
        self._buffer = ""
        self._current = SseEvent()

    def push_chunk(self, chunk: bytes) -> list[SseEvent]:
        """Decode a raw chunk, returning every event completed by it."""
        try:
            text = chunk.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ValueError(f"SSE chunk is not valid UTF-8: {exc}") from exc
        self._buffer += text

        events: list[SseEvent] = []
        while True:
            index = self._buffer.find("\n")
            if index == -1:
                break
            line = self._buffer[:index]
            self._buffer = self._buffer[index + 1 :]
            if line.endswith("\r"):
                line = line[:-1]
            event = self._process_line(line)
            if event is not None:
                events.append(event)
        return events

    def finish(self) -> list[SseEvent]:
        """Flush an incomplete final event, if any, at EOF."""
        events: list[SseEvent] = []
        if self._buffer:
            line = self._buffer
            self._buffer = ""
            event = self._process_line(line.rstrip("\r"))
            if event is not None:
                events.append(event)
        event = self._dispatch_current()
        if event is not None:
            events.append(event)
        return events

    def _process_line(self, line: str) -> SseEvent | None:
        if not line:
            return self._dispatch_current()
        if line.startswith(":"):
            return None

        if ":" in line:
            field_name, value = line.split(":", 1)
            if value.startswith(" "):
                value = value[1:]
        else:
            field_name, value = line, ""

        if field_name == "event":
            self._current.event = value
        elif field_name == "data":
            if self._current.data:
                self._current.data += "\n"
            self._current.data += value
        elif field_name == "id":
            self._current.id = value
        elif field_name == "retry":
            with contextlib.suppress(ValueError):
                self._current.retry = int(value, 10)
        return None

    def _dispatch_current(self) -> SseEvent | None:
        current = self._current
        if current.event is None and not current.data and current.id is None and current.retry is None:
            return None
        self._current = SseEvent()
        return current


@dataclass
class MeteredSseEvent:
    """A parsed metered SSE event, tagged by ``kind``.

    Exactly one payload field is set for the matching ``kind``; a ``done`` event
    carries no payload.
    """

    #: Which kind of event this is, selecting the populated payload field.
    kind: Literal["metering", "usage", "message", "done", "other"]
    #: Metering directive payload, set when ``kind`` is ``"metering"``.
    metering: MeteringDirective | None = None
    #: Usage payload, set when ``kind`` is ``"usage"``.
    usage: MeteringUsage | None = None
    #: Decoded JSON message payload, set when ``kind`` is ``"message"``.
    message: Any = None
    #: Raw passthrough frame, set when ``kind`` is ``"other"``.
    other: SseEvent | None = None


def parse_metered_sse_event(event: SseEvent) -> MeteredSseEvent:
    """Classify an SSE frame as metering, usage, message, done, or other.

    Recognizes the canonical event names ``mpp.metering``/``metering`` and
    ``mpp.usage``/``usage``, the ``done`` event, and the ``[DONE]`` sentinel on
    a plain message. Message payloads are decoded as JSON values; any other
    event name passes through unchanged as an ``"other"`` frame.
    """
    event_name = event.event if event.event is not None else "message"
    if event_name in ("mpp.metering", "metering"):
        try:
            directive = MeteringDirective.from_dict(json.loads(event.data))
        except (ValueError, TypeError, AttributeError) as exc:
            raise ValueError(f"invalid mpp.metering event: {exc}") from exc
        return MeteredSseEvent(kind="metering", metering=directive)
    if event_name in ("mpp.usage", "usage"):
        try:
            usage = MeteringUsage.from_dict(json.loads(event.data))
        except (ValueError, TypeError, AttributeError) as exc:
            raise ValueError(f"invalid mpp.usage event: {exc}") from exc
        return MeteredSseEvent(kind="usage", usage=usage)
    if event_name == "done":
        return MeteredSseEvent(kind="done")
    if event_name == "message":
        if event.data.strip() == "[DONE]":
            return MeteredSseEvent(kind="done")
        try:
            message = json.loads(event.data)
        except ValueError as exc:
            raise ValueError(f"invalid SSE message event: {exc}") from exc
        return MeteredSseEvent(kind="message", message=message)
    return MeteredSseEvent(kind="other", other=event)


class _MeteredStreamState:
    """Directive/usage/done bookkeeping shared by the stream wrappers.

    Tracks the live metering directive, the final usage amount once one
    arrives, and whether the stream has ended.
    """

    def __init__(self) -> None:
        self.directive: MeteringDirective | None = None
        self.final_amount: int | None = None
        self.done = False

    def apply_event(self, event: SseEvent) -> Any | None:
        """Fold one SSE frame into the state, returning a message payload."""
        parsed = parse_metered_sse_event(event)
        if parsed.kind == "metering":
            self.directive = parsed.metering
            return None
        if parsed.kind == "usage":
            usage = parsed.usage
            if usage is None:  # pragma: no cover - parse always sets it
                raise ValueError("invalid mpp.usage event: missing payload")
            if self.directive is not None and usage.delivery_id != self.directive.delivery_id:
                raise ValueError(
                    f"usage delivery {usage.delivery_id} does not match directive {self.directive.delivery_id}"
                )
            self.final_amount = usage.amount_base_units()
            return None
        if parsed.kind == "message":
            return parsed.message
        if parsed.kind == "done":
            self.done = True
        return None

    def directive_for_commit(self) -> MeteringDirective:
        """Return the directive to commit, with usage overriding the amount.

        Usage overrides only the amount, never the ``deliveryId``.
        """
        if self.directive is None:
            raise ValueError("stream did not include mpp.metering event")
        if self.final_amount is not None:
            return replace(self.directive, amount=str(self.final_amount))
        return self.directive


class MeteredSseSession:
    """Borrowed state machine for a metered SSE stream.

    Feed decoded frames with :meth:`accept_event`; call :meth:`ack` after the
    stream ends to sign and commit a voucher for the metered amount (the
    reserved directive amount, or the final usage amount when one arrived).
    This wrapper does not own its frame source: the caller decodes and feeds
    frames in.
    """

    def __init__(self, consumer: SessionConsumer) -> None:
        """Wrap a session consumer with fresh metering state."""
        self._consumer = consumer
        self._state = _MeteredStreamState()

    @property
    def consumer(self) -> SessionConsumer:
        """The wrapped consumer."""
        return self._consumer

    def accept_event(self, event: SseEvent) -> Any | None:
        """Fold one SSE frame in, returning a message payload when present."""
        return self._state.apply_event(event)

    @property
    def is_done(self) -> bool:
        """True once a ``done`` event or ``[DONE]`` sentinel arrived."""
        return self._state.done

    def ack(self) -> CommitReceipt:
        """Sign and commit a voucher for the streamed delivery."""
        directive = self._state.directive_for_commit()
        return self._consumer.commit_directive(directive)


class MeteredSseStream:
    """A metered SSE stream over any iterable of raw byte chunks.

    Works with ``httpx.Response.iter_bytes()`` or any other source of raw byte
    chunks; the stream stays transport-neutral and owns its own decoder and
    metering state. Iterate with :meth:`next` (or as an iterator) and call
    :meth:`ack` once to commit.
    """

    def __init__(self, consumer: SessionConsumer, chunks: Iterable[bytes]) -> None:
        """Wrap a consumer and a raw chunk source."""
        self._consumer = consumer
        self._chunks = iter(chunks)
        self._decoder = SseDecoder()
        self._pending: list[Any] = []
        self._state = _MeteredStreamState()

    def next(self) -> Any | None:
        """Return the next message payload, or ``None`` at end of stream."""
        while True:
            if self._pending:
                return self._pending.pop(0)
            if self._state.done:
                return None
            chunk = next(self._chunks, None)
            if chunk is None:
                for event in self._decoder.finish():
                    message = self._state.apply_event(event)
                    if message is not None:
                        self._pending.append(message)
                self._state.done = True
                continue
            for event in self._decoder.push_chunk(chunk):
                message = self._state.apply_event(event)
                if message is not None:
                    self._pending.append(message)

    def __iter__(self) -> Iterator[Any]:
        """Yield message payloads until the stream ends."""
        while True:
            message = self.next()
            if message is None:
                return
            yield message

    def ack(self) -> CommitReceipt:
        """Drain the stream if needed, then sign and commit the delivery."""
        if not self._state.done:
            while self.next() is not None:
                pass
        directive = self._state.directive_for_commit()
        return self._consumer.commit_directive(directive)

    def into_consumer(self) -> SessionConsumer:
        """Return the wrapped consumer for reuse on the next request."""
        return self._consumer


class HttpCommitTransport:
    """Minimal HTTP transport for commit endpoints.

    Posts the :class:`CommitPayload` JSON to the directive ``commitUrl`` (or a
    default), optionally with an Authorization header, and decodes the
    :class:`CommitReceipt`. Pass an ``httpx.Client`` to control timeouts or to
    inject a mock transport in tests; one is created lazily when omitted.
    """

    def __init__(
        self,
        client: Any | None = None,
        default_commit_url: str | None = None,
        authorization: str | None = None,
    ) -> None:
        """Create a transport over an optional httpx client and defaults."""
        if client is None:
            import httpx

            client = httpx.Client()
        self._client = client
        self._default_commit_url = default_commit_url
        self._authorization = authorization

    def commit(self, directive: MeteringDirective, payload: CommitPayload) -> CommitReceipt:
        """POST the commit payload and decode the receipt."""
        url = directive.commit_url if directive.commit_url is not None else self._default_commit_url
        if url is None:
            raise ValueError("metering directive missing commitUrl")

        headers: dict[str, str] = {}
        if self._authorization is not None:
            headers["authorization"] = self._authorization
        try:
            response = self._client.post(url, json=payload.to_dict(), headers=headers)
        except Exception as exc:  # noqa: BLE001 - transport errors surface as commit failures
            raise ValueError(f"commit request failed: {exc}") from exc
        if response.status_code < 200 or response.status_code >= 300:
            raise ValueError(f"commit endpoint returned {response.status_code}: {response.text}")
        try:
            return CommitReceipt.from_dict(response.json())
        except (ValueError, TypeError) as exc:
            raise ValueError(f"invalid commit receipt: {exc}") from exc
